src/main/java/com/limechain/runtime/hostapi/CryptoHostFunctions.java

```diff
@@ -1,99 +1,477 @@
+                        List.of(Type.I32, Type.I32), (Type.I64)),
+                HostApi.getImportObject("ext_crypto_secp256k1_ecdsa_recover_version_2", argv ->
+                                secp256k1EcdsaRecoverV1((int) argv.get(0), (int) argv.get(1)),
+                        List.of(Type.I32, Type.I32), Type.I64),
                 HostApi.getImportObject("ext_crypto_secp256k1_ecdsa_recover_compressed_version_1",
-                        argv -> {
-                            return argv;
-                        }, List.of(Type.I32, Type.I32), Type.I64),
+                        argv -> secp256k1EcdsaRecoverCompressedV1((int) argv.get(0), (int) argv.get(1)),
+                        List.of(Type.I32, Type.I32), Type.I64),
                 HostApi.getImportObject("ext_crypto_secp256k1_ecdsa_recover_compressed_version_2",
-                        argv -> {
-                            return argv;
-                        }, List.of(Type.I32, Type.I32), Type.I64),
-                HostApi.getImportObject("ext_crypto_start_batch_verify", argv -> {
-
-                }, HostApi.EMPTY_LIST_OF_TYPES),
-                HostApi.getImportObject("ext_crypto_finish_batch_verify", argv -> {
-                    return argv;
-                }, HostApi.EMPTY_LIST_OF_TYPES, Type.I32));
+                        argv -> secp256k1EcdsaRecoverCompressedV1((int) argv.get(0), (int) argv.get(1)),
+                        List.of(Type.I32, Type.I32), Type.I64),
+                HostApi.getImportObject("ext_crypto_start_batch_verify", argv ->
+                        startBatchVerify(), HostApi.EMPTY_LIST_OF_TYPES),
+                HostApi.getImportObject("ext_crypto_finish_batch_verify", argv ->
+                        finishBatchVerify(), HostApi.EMPTY_LIST_OF_TYPES, Type.I32));
+    }
+
+    private VerifySignature internalGetVerifySignature(int signature, long message, int publicKey, Key key) {
+        final byte[] signatureData = hostApi.getDataFromMemory(signature, 64);
+        final byte[] messageData = hostApi.getDataFromMemory(new RuntimePointerSize(message));
+        final byte[] publicKeyData = hostApi.getDataFromMemory(publicKey, 32);
+        return new VerifySignature(signatureData, messageData, publicKeyData, key);
+    }
+
+    private int ed25519PublicKeysV1(int keyTypeId) {
+        final KeyType keyType = KeyType.getByBytes(hostApi.getDataFromMemory(keyTypeId, 4));
+
+        if (keyType == null || keyType.getKey() != Key.ED25519 && keyType.getKey() != Key.GENERIC) {
+            //Todo: How to handle exceptions?
+            return -1;
+        }
+
+        byte[] publicKeys = keyStore.getPublicKeysByKeyType(keyType);
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        ScaleCodecWriter scaleWriter = new ScaleCodecWriter(baos);
+        try {
+            scaleWriter.writeAsList(publicKeys);
+        } catch (IOException e) {
+            //Todo: How to handle exceptions?
+            return -1;
+        }
```

Comment:
Define a byte array before the try and use [try with resources](https://www.baeldung.com/java-try-with-resources) for creating the stream and writer to have them autoclose at the end.

Reply:
This goes for all classes that implement `Closable`